### DISCUSSION TITLE
RedisCodec Implementation Question

### DISCUSSION THREAD

**Discussion by dalegaspi** [redis/lettuce#1864] [Q&A] [answered]
Hello!  I have a question regarding implementing a custom codec.

I am using lettuce to read a gzip-compressed string in Redis and deserializing the said string in a Scala application using `json4s::parse` (with Jackson).

Currently we are reading the gzip data into a String then call `json4s::parse`.  I thought the conversion to String is not necessary (and waste of memory) since `json4s::parse` already takes in an `InputStream` parameter.

Now I have implemented a custom codec (for reads only) that returns an `InputStream` 

```java
public class StringInputStreamCodec implements RedisCodec<String, InputStream> {
    private final Charset charset = StandardCharsets.UTF_8;

    @Override
    public String decodeKey(final ByteBuffer bytes) {
        return charset.decode(bytes).toString();
    }

    @Override
    public InputStream decodeValue(final ByteBuffer bytes) {
        return new ByteBufferBackedInputStream(bytes);
    }

    @Override
    public ByteBuffer encodeKey(final String key) {
        return charset.encode(key);
    }

    @Override
    public ByteBuffer encodeValue(final InputStream value) {
        throw new NotImplementedException("This is a read-only codec!");
    }
}
```

And the `ByteBufferBackedInputStream` was an implementation from Jackson's [`com.fasterxml.jackson.databind.util.ByteBufferBackedInputStream`](https://fasterxml.github.io/jackson-databind/javadoc/2.8/com/fasterxml/jackson/databind/util/ByteBufferBackedInputStream.html#ByteBufferBackedInputStream(java.nio.ByteBuffer))

Usage:

```scala
val conn = client.connect(CompressionCodec.valueCompressor(
    new StringInputStreamCodec(), CompressionCodec.CompressionType.GZIP)).async();
```

The question: my tests shows that it is working but...is this a sound approach?  I'm reading [here](https://github.com/lettuce-io/lettuce-core/wiki/Codecs) about the Codecs being thread-safe but it didn't warn anything about NOT referencing the underlying implementation of the data returned as `ByteBuffer` so I figured it should be safe to use directly by "wrapping" it with an InputStream reader?

Thanks!

**Comment by mp911de** [comment, marked as answer]
I converted the issue into a discussion. 

The `ByteBuffer` interface in codecs is used by Lettuce accepting and providing buffers. When decoding, you should either copy or consume buffers as the input buffer is attached to a Netty buffer. Keeping a reference may result in a changed underlying buffer and undesired consequences.

**Reply by dalegaspi** [reply]
ok.  that makes sense.   thanks, @mp911de.
